nuclear: 0.6.6 -> unstable-2022-04-20

WolfangAukang opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

Latest stable version (0.6.17) does not work, but one of the latest pre-releases works fine. Also, as this app is still an alpha, using an unstable version shouldn't be a problem.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

IvarWithoutBones commented:
I'm getting the following error when running this from nixpkgs-review:
```console
[nix-shell:~/.cache/nixpkgs-review/pr-169347-1]$ ./results/nuclear/bin/nuclear
      main › (node:2) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
(Use `nuclear --trace-deprecation ...` to show where the warning was created)
[2:0421/193633.580170:FATAL:gpu_data_manager_impl_private.cc(415)] GPU process isn't usable. Goodbye.
```
Adding the `--no-sandbox` flags fixes it, although that's a bit annoying. According to https://github.com/LedgerHQ/ledger-live-desktop/issues/4253#issuecomment-926215043 this should be fixed with an upstream electron update, we might want to wrap the binary in the meantime.

Other than that, this looks good to me :+1:

WolfangAukang commented:
> we might want to wrap the binary in the meantime
.
Sure thing, just wrapped it

IvarWithoutBones commented on pkgs/applications/audio/nuclear/default.nix:
```diff
@@ -17,8 +19,11 @@ in appimageTools.wrapType2 {
+    # Adding --no-sandbox as a temporary fix https://github.com/NixOS/nixpkgs/pull/169347#issuecomment-1105507295
     substituteInPlace $out/share/applications/${pname}.desktop \
-      --replace 'Exec=AppRun' 'Exec=${pname}'
+      --replace 'Exec=AppRun' 'Exec=${pname} --no-sandbox'
```
This way the fix only applies to the desktop file, so launching the app from the terminal would still cause issues. You should use `wrapProgram` instead:

```suggestion
      --replace 'Exec=AppRun' 'Exec=${pname}'

      # Temporary fix for https://github.com/nukeop/nuclear/issues/1268
      wrapProgram $out/bin/nuclear --add-flags "--no-sandbox"
```

WolfangAukang replied:
Here is the thing: To do this, you need to run wrapProgram into a `nativeBuildInputs` list so it can be recognized for its usage on the phase. As this is an AppImage, I'm not sure of the correct approach, as `nativeBuildInputs` is an anonymous function at this scope. That's why I only wrapped the desktop file.

WolfangAukang replied:
Ok, I was able to move it into an `mkDerivation`, but some considerable observations:

- As we are just wrapping to wait for a fix, it would be preferred to revert it into the old structure (the AppImage wrapType2) after the fix is finally applied.
- Nuclear maximum supported Electron version is 13, which is currently marked as insecure.

IvarWithoutBones replied:
Mhmm, that does seem a bit annoying. I saw the other PR that updates this used `extraPkgs = pkgs: with pkgs; [ wrapGAppsHook ];`, which would only do anything if it's an `nativeBuildInput` right? Wouldn't that work the same with `makeWrapper`?

IvarWithoutBones reviewed (commented):
I have created an issue upstream so that we can hopefully get this fixed in a future release. See https://github.com/nukeop/nuclear/issues/1268.

WolfangAukang committed:
nuclear: 0.6.6 -> unstable-2022-04-20

Using stdenv.mkDerivation instead of appimageTools.wrapType2 as a temporary workaround because the binary needs to be wrapped with the --no-sandbox flag

WolfangAukang commented:
Moving to [NUR](https://nur.nix-community.org/repos/wolfangaukang/)
